openmmapi/include/openmm/NoseHooverIntegrator.h

```diff
@@ -0,0 +1,279 @@
+     *
+     * @param chainID the index of the Nose-Hoover chain thermostat (default=0).
+     *
+     * @return the collision frequency.
+     */
+    double getRelativeCollisionFrequency(int chainID=0) const;
+    /**
+     * Set the collision frequency for pairs' relative motion of the i-th chain.
+     *
+     * @param frequency the collision frequency in picosecond.
+     * @param chainID the index of the Nose-Hoover chain thermostat (default=0).
+     */
+    void setRelativeCollisionFrequency(double frequency, int chainID=0);
+    /**
+     * Compute the total (potential + kinetic) heat bath energy for all heat baths
+     * associated with this integrator, at the current time.
+     */
+    double computeHeatBathEnergy();
+    /**
+     * Get the number of Nose-Hoover chains registered with this integrator.
+     */
+    int getNumNoseHooverThermostats() const {
+        return noseHooverChains.size();
+    }
+    /**
+     * Get the NoseHooverChain thermostat 
+     *
+     * @param chainID the index of the Nose-Hoover chain thermostat (default=0).
+     */
+    const NoseHooverChain& getNoseHooverThermostat(int chainID=0) const ;
+    /**
+     * This will be called by the Context when the user modifies aspects of the context state, such
+     * as positions, velocities, or parameters.  This gives the Integrator a chance to discard cached
+     * information.  This is <i>only</i> called when the user modifies information using methods of the Context
+     * object.  It is <i>not</i> called when a ForceImpl object modifies state information in its updateContextState()
+     * method (unless the ForceImpl calls a Context method to perform the modification).
+     * 
+     * @param changed     this specifies what aspect of the Context was changed
+     */
+    virtual void stateChanged(State::DataType changed) {
+       if (State::Positions == changed) forcesAreValid = false;
+    }
+    /**
+     * Return false, if this integrator was set up with the 'default constructor' that thermostats the whole system,
+     * true otherwise. Required for serialization.
+     */
+    bool hasSubsystemThermostats() const {
+        return hasSubsystemThermostats_;
+    }
+    /**
+     * Gets the maximum distance (in nm) that a connected pair may stray from each other. If zero, there are no
+     * constraints on the intra-pair separation.
+     */
+    double getMaximumPairDistance() const { return maxPairDistance_; }
+    /**
+     * Sets the maximum distance (in nm) that a connected pair may stray from each other, implemented using a hard
+     * wall. If set to zero, the hard wall constraint is omited and the pairs are free to be separated by any distance.
+     */
+    void setMaximumPairDistance(double distance) { maxPairDistance_ = distance; }
+    /**
+     * Get a list of all individual atoms (i.e. not involved in a connected Drude-like pair) in the system.
+     */
+    const std::vector<int> & getAllAtoms() const { return allAtoms; }
```

Comment:
I'm a little confused about this method, both what it's for and the intended behavior.  If it omits any particles that are part of a pair, then it isn't really "all atoms", is it?  Also, if the thermostat is only applied to a subset of particles, should it still return all particles, or only the thermostated ones?  I'm just not clear what this is for.

Public parts of the C++ API should use the word "particles" rather than "atoms".

Reply:
You are right, this name is confusing. We can change it to `getThermostatedIndividualParticles` and `getThermostatedPairs` or something like that.